feat: add react and typescript standards for performance (#1)
automation-consulting-2024/coding-standards-and-conventions: react/README.md

# General

## Basic Component Structure

> - Simple Component Structure

Reviewer: I think the example will make more sense if we accompany it with explanations why each sections are in the position they are in.

```jsx

> function Component() {

Reviewer: ```suggestion

interface ComplextState {
  isState: boolean;
}

const Component = () => {
```

    // MARK: - States

>     const [state, setState] = useState()

Reviewer: ```suggestion
    const [state, setState] = useState<ComplexState | null>(null)
```

>     // MARK: - Values mapped from States

Reviewer: What are these `// MARK`?

    const isState = state.isState

    // MARK: - Event Handlers
    const onClick = () => {
        return onClick()

>     }

Reviewer: This is confusing. Are you recursively calling `onClick`?

Don't want to nitpick but this is going to be passed to an LLM so we need to make sure things are correct.

    // MARK: - Side Effecs
    useEffect(() => {}, [
        setState(true)

>     ])

Reviewer: Hmmm... why is `setState` in dependencies array?

    // MARK: - Renderer
    return (
        <>
            {content..}
        </>
    )
}
```

## Performance

> ### Prefering Pure-Components

Reviewer: ```suggestion
### Preferring Pure-Components
```

- Based on the concept of purity in functional programming paradigms, a function is said to be pure if it meets the following two conditions:
  - Its return value is only determined by its input values

>   - Its return value is always the same for the same input values

Reviewer: These feel like the same point but written in different ways.

I think the other point for pure functions are those without side effects.

```jsx

> function PercentageStat({ label, score = 0, total = Math.max(1, score) }) {

Reviewer: I prefer arrow functions for components. Thoughts?

Also since we will most likely use typescript, I think the examples here should be `tsx`.

  return (
    <div>
      <h6>{label}</h6>
      <span>{Math.round((score / total) * 100)}%</span>
    </div>
  );
}
```

### Move Constants & Pure function to outside component

> - Everything inside a component have to be re-initialize every time it re-render, move
>   them to outside of component help JS to only initialize them one.

Reviewer: ```suggestion
- Everything inside a component have to be re-initialize every time it re-renders. For functions and variables that do not require states or setters internally, it is recommended to move them to outside of component. In this way, Javascript only initializes them once.
```

```jsx
const DEFAULT_MULTIPLY = 2;

> const multiplyBy = (a, b) => {

Reviewer: ```suggestion
const multiplyBy = (a: number, b: number): number => {
```

  return a + b * DEFAULT_MULTIPLY;
};

const Component = ({ a, b }) => {

>   const c = add(a + b);

Reviewer: ```suggestion
  const c = multiply(a + b);
```

  return <>{c}</>;
};
```

### Initialize expensive state by arrow function

```jsx
// instead of this which would be executed on every re-render:
const [state, setState] = React.useState(myExpensiveFn());

// prefer this which is executed only once:
const [state, setState] = React.useState(() => myExpensiveFn());
```

### Before you Memo

- Not every performance problem end with `useMemo` or `React.memo`.

Here a `slow` React component:

```jsx
import { useState } from 'react';

export default function App() {

>   let [color, setColor] = useState('red');

Reviewer: ```suggestion
  const [color, setColor] = useState('red');
```

  return (
    <div>
      <input value={color} onChange={(e) => setColor(e.target.value)} />
      <p style={{ color }}>Hello, world!</p>
      <ExpensiveTree />
    </div>
  );
}

function ExpensiveTree() {
  let now = performance.now();
  while (performance.now() - now < 100) {
    // Artificial delay -- do nothing for 100ms
  }
  return <p>I am a very slow component tree.</p>;

> 

Reviewer 2: ```suggestion
}

```

Problem: Everytime `color` state change, `ExpensiveTree` will re-render.

#### Solution 1: Move State Down

- As you see above, only `input` and `p` element care about `color`
  state, so we should able to extract relative logic to they own component
  and move the state down like bellow:

```jsx
export default function App() {
  return (
    <>
      <Form />
      <ExpensiveTree />
    </>
  );
}

function Form() {

>   let [color, setColor] = useState('red');

Reviewer: ```suggestion
  const [color, setColor] = useState('red');
```

  return (
    <>
      <input value={color} onChange={(e) => setColor(e.target.value)} />
      <p style={{ color }}>Hello, world!</p>
    </>
  );
}
```

> #### Solution 2: Lift content up

Reviewer: Could you potentially use memoization as solution #3?
